README.md
# clj-gearman

This library provides Clojure bindings for [gearman](http://gearman.org/) distributed job system.

[You can find the protocol specification here.](http://gearman.org/protocol/)

This implementation uses regular blocking sockets.
Take a look at [clearman](https://github.com/joshrotenberg/clearman) if you need async IO.

# Installation

Add the following dependency to your project.clj file:

[![Clojars Project](https://clojars.org/clj-gearman/latest-version.svg)](https://clojars.org/clj-gearman)

# Synopsis

Worker side:

```clojure
(ns my-worker
  (:require [clojure.string :as str]
            [clj-gearman.worker :as w]))

(def worker

   ; List of job servers you have running.
  {:job-servers [{:host "localhost" :port 4730}]

   ; Number of threads to run per job server in a worker pool.
   ; This will be multiplied by the number of job servers.
   ; For example, if you connect to three servers and :nthreads if five,
   ; you will get 15 worker threads in total.
   ;
   ; Defaults to one.
   :nthreads 4

   ; String encoding for input and output.
   ; You probably want to keep these utf-8.
   :in-enc  "UTF-8"
   :out-enc "UTF-8"

   ; Map of your worker capability functions.
   ; Notice that keys should be strings.
   :can-do {
            ; First is the ever popular string reversal.
            ; Workload parameter is always a string and your return value must
            ; also be a single string.
            "reverse" (fn [workload] (str/reverse workload))

            ; If you like a more hands-on approach you can operate on the worker socket directly.
            ; In this case you are responsible for all communication with the job server.
            "long-running" (fn [socket job-handle workload]
                              (try
                               (do
                                 (your-slow-operation-one workload)
                                 (w/work-status socket job-handle 1 3)
                                 (your-slow-operation-two workload)
                                 (w/work-status socket job-handle 2 3)
                                 (your-slow-operation-three workload)
                                 (w/work-status socket job-handle 3 3)
                               (catch Throwable ex
                                 (w/work-exception socket job-handle ex))))}})

; Start a worker pool.
(def pool-fn (w/pool worker))

; When you are done, stop workers by calling the function returned previously.
; Note that this might take some time, especially if there is a long running task in progress.
(pool-fn)


; Alternatively, if you just need to run a single worker on a single server
; you can use this lower-level method directly. Or you can use this to create your own
; worker pool implementation.
;
; Connect to job server and start accepting tasks.
(with-open [socket (w/connect worker)]
  (while true (w/work socket)))

```


Client side:

```clojure
(ns my-client
  (:require [clojure.pprint :refer [pprint]]
            [clj-gearman.client :as c]))

(def client

   ; List of job servers you have running.
   ; Client connects to first server available.
  {:job-servers [{:host "localhost" :port 4730}]

   ; String encoding for input and output.
   ; You probably want to keep these utf-8.
   :in-enc  "UTF-8"
   :out-enc "UTF-8"

   ; If true, exceptions thrown in worker will be propagated through
   ; the job server to this client.
   :worker-exceptions false})

; Connect to job server and submit the work requests.
(with-open [socket (c/connect client)]
  (let [[code result :as response] (c/submit-job socket "reverse" "foo bar baz")]
    (if (= code "WORK_COMPLETE")
      (println result)
      (throw (Exception. (pr-str response)))))

  ; Fire-and-forget background job.
  ; Use this if you don't need the final return value from the worker.
  (let [[code job-handle :as response] (c/submit-job-bg socket "long-running" "Our workload string")]
    (if (not= code "JOB_CREATED")
      ; Oops, that's an error.
      (throw (Exception. (pr-str response)))
      ; Otherwise, we are done. Job is sent to job server and hopefully some worker will pick it up.
      ; Worker will not send us the result but we can optionally poll for status of the task.
      (comment
        (dotimes [_ 10]
          (Thread/sleep 1000)
          (pprint (c/get-status socket job-handle)))))))

```

# API documentation

[Read the full api docs here.](https://ajk.github.io/clj-gearman/index.html)


# Contributing

Feel free to send bug reports and pull requests on this repository.

# License

Copyright 2016 Antti Koskinen

Distributed under the Eclipse Public License, the same as Clojure.


CHANGELOG.md
0.5.1 (2016-12-19)

- Worker pool implementation.

0.4.0 (2016-11-07)

- Support for multiple Gearman servers.

0.3.0 (2016-10-23)

- API documentation, more consistent parameter naming, refactoring.

0.2.0 (2016-10-22)

- Added Socket exception handling.

0.1.0 (2016-10-19)

- Initial release.
- First draft of the Gearman client and worker implementation.


test/clj_gearman/util_test.clj
(ns clj-gearman.util-test
  (:require [clojure.test :refer :all]
            [clj-gearman.util :as u]))


(deftest convert-ints
  (testing "Convert integers to byte arrays and back."
    (is (= 0 (u/bytea->int (u/int->bytea 0))))
    (is (= 1 (u/bytea->int (u/int->bytea 1))))
    (is (= -123 (u/bytea->int (u/int->bytea -123))))
    (is (= 1024 (u/bytea->int (u/int->bytea 1024))))
    (is (= 2147483647 (u/bytea->int (u/int->bytea 2147483647))))
    (is (thrown? java.lang.IllegalArgumentException (u/int->bytea 2147483648)))))

(deftest convert-strings
  (testing "Convert strings to byte arrays and back."
    (is (= "" (u/bytea->str (u/str->bytea ""))))
    (is (= "abc..åäö" (u/bytea->str (u/str->bytea "abc..åäö"))))
    (is (= "abc" (u/bytea->str (u/str->bytea "abc" "ASCII") "ASCII")))
    (is (not= "åäö" (u/bytea->str (u/str->bytea "åäö") "ASCII")))
    (is (= "åäö" (u/bytea->str (u/str->bytea "åäö" nil) "UTF-8")))
    (is (= "åäö" (u/bytea->str (u/str->bytea "åäö" "UTF-16") "UTF-16")))
    (is (thrown? java.io.UnsupportedEncodingException (u/str->bytea "foo" "FOO")))))

; For testing purposes, flatten our null-separated byte array

(deftest convert-msg
  (testing "Convert message payload to null-separated byte arrays and back"
    (is (= []                  (-> []                  (u/msg->bytea "ASCII") (u/bytea->msg "ASCII"))))
    (is (= ["echo"]            (-> ["echo"]            (u/msg->bytea "ASCII") (u/bytea->msg "ASCII"))))
    (is (= ["Foo" "Bar" "Bäz"] (-> ["Foo" "Bar" "Bäz"] (u/msg->bytea "UTF-8") (u/bytea->msg "UTF-8"))))))


test/clj_gearman/client_test.clj
(ns clj-gearman.client-test
  (:require [clojure.test :refer :all]
            [clj-gearman.client :as c]))

; NOTICE: These test will fail unless gearman server is running in port 5555

; First job server doesn't exist but connection to second one should succeed.
(def my-client {:job-servers [{:host "localhost" :port 5554}
                              {:host "localhost" :port 5555}]})

(deftest echo
  (testing "Echo request and response"
    (with-open [client (c/connect my-client)]
      (is (= ["ECHO_RES" "echö"] (c/echo client "echö"))))))


test/clj_gearman/socket_test.clj
(ns clj-gearman.socket-test
  (:require [clojure.test :refer :all]
            [clj-gearman.socket :as s]
            [clj-gearman.header :as h]))

; NOTICE: These test will fail unless gearman server is running in port 5555

(defn echo-req [data test-fn]
  (let [sock (s/connect {:host "localhost" :port 5555})]
    (s/write-msg sock h/req (h/code "ECHO_REQ") "UTF-8" [data])
    (test-fn (s/read-msg sock "UTF-8"))
    (s/disconnect sock)))

(deftest raw-echo
  (testing "Send a byte array to gearman server, expect echo response."
    (let [sock (s/connect {:host "localhost" :port 5555})
          out (.getOutputStream @sock)
          in  (.getInputStream @sock)
          bytes-out (byte-array '(0, 82, 69, 81, 0, 0, 0, 16, 0, 0, 0, 4, 101, 99, 104, 111))
          bytes-in  (byte-array '(0, 82, 69, 83, 0, 0, 0, 17, 0, 0, 0, 4, 101, 99, 104, 111))]
      (.write out bytes-out)
      (.flush out)
      (is (= (vec bytes-in) (vec (s/read-bytes in (count bytes-out)))))
      (s/disconnect sock))))

(deftest echo
  (testing "Echo request and response"
    (echo-req "echö"
              (fn [[type code size data]]
                (is (= (vec type) (vec h/res)))
                (is (= code (h/code "ECHO_RES")))
                (is (= 5 size))
                (is (= "echö" (last data)))))))


src/clj_gearman/pool.clj
(ns clj-gearman.pool
  "Gearman job server connection pool functions."
  (:import (java.net ConnectException))
  (:require [clj-gearman.socket :as s]))

(defn- conn-socket [meta retry]
  (fn [conn] (try (s/connect conn meta retry) (catch ConnectException ex nil))))

(defn- conn-slow [servers meta]
  (if-let [socket (some (conn-socket meta 10) (butlast servers))]
    socket
    (s/connect (last servers) meta 10)))

(defn- conn-fast [servers meta]
  (some (conn-socket meta 0) servers))

(defn connect-first
  "Loop through a list of servers and return first one available"
  [servers conn-meta]
  ; First try each server without retries.
  (if-let [socket (conn-fast servers conn-meta)]
    socket
    ; If that fails, loop them again and throw exception if we still can't connect.
    (conn-slow servers conn-meta)))


(defn- worker-thread [running worker connect work]
  (Thread. (fn []
             (while @running
               (try
                 (with-open [socket (connect worker)]
                   (while @running (work socket)))
                 (catch Throwable _ (Thread/sleep 5000)))))))


(defn worker-pool [worker-def connect work]
  (let [running (atom true)
        workers (map #(assoc worker-def :job-servers [%]) (:job-servers worker-def))
        server-pool (fn [worker]
                      (map #(worker-thread running worker connect work)
                           (range (get worker-def :nthreads 1))))
        server-pools (map server-pool workers)]
    (doseq [pool server-pool]
      (doseq [^Thread th pool]
        (.start th)))
    (fn []
      (reset! running false)
      (doseq [pool server-pool]
        (doseq [^Thread th pool]
          (.join th))))))


src/clj_gearman/util.clj
(ns clj-gearman.util
  "Miscellaneous utility functions."
  (:import (java.nio ByteBuffer))
  (:require [clojure.pprint :refer [pprint]]))

(defn new-uniq
  "Generates a new version 4 UUID."
  []
  (str (java.util.UUID/randomUUID)))

(defn fun-arity
  "Returns the count of arguments that the function takes."
  [fun]
  (->> (class fun)
       .getDeclaredMethods
       (filter #(= "invoke" (.getName %)))
       first
       .getParameterTypes
       alength))

(defn int->bytea
  "Converts signed integer to a byte array in big-endian order."
  [^Long x]
  (-> (ByteBuffer/allocate 4)
      (.putInt x)
      (.array)))

(defn bytea->int
  "Converts a byte array in big-endian order to signed integer."
  [^bytes byte-a]
  (-> (ByteBuffer/allocate 4)
      (.put byte-a)
      (.getInt 0)))

(defn str->bytea
  "Converts a string in specific encoding to a byte array. Defaults to UTF-8."
  ([str] (str->bytea str "UTF-8"))
  ([^String str ^String enc]
   (.getBytes str (or enc "UTF-8"))))

(defn bytea->str
  "Converts a byte array to a string in specific encoding. Defaults to UTF-8."
  ([byte-a] (bytea->str byte-a "UTF-8"))
  ([^bytes byte-a ^String enc]
   (String. byte-a (or enc "UTF-8"))))


(defn concat-bytea
  "Concatenates a sequence of byte arrays into a single one."
  ^bytes
  [byte-arrays]
  (let  [size (reduce + (map count byte-arrays))
         ^bytes out (byte-array size)
         bb (ByteBuffer/wrap out)]
    (doseq [^bytes ba byte-arrays]
      (.put bb ba))
    out))

(defn msg->bytea
  "Converts a sequence of strings to a null-separated byte array."
  [msg enc]
  (if (empty? msg)
    (byte-array 0)
    (let [handle-args (mapv #(str->bytea (str %) "ASCII") (or (butlast msg) ()))
          data-arg    (str->bytea (str (last msg)) enc)]
      (->> data-arg
           (conj handle-args)
           (interpose (byte-array [(byte 0)]))
           concat-bytea))))

(defn- split-null
  "Splits byte-array to chunks separated by null bytes"
  [byte-a]
  (map byte-array (filter #(or (> (count %) 1) (not= (first %) (byte 0)))
                          (partition-by zero? byte-a))))

(defn bytea->msg
  "Converts a null-separated byte array to a sequence of strings."
  [byte-a enc]
  (if-not (empty? byte-a)
    (let [chunks (split-null byte-a)
          handle-args (if (< (count chunks) 2) [] (mapv #(bytea->str % "ASCII") (butlast chunks)))
          data-arg    (bytea->str (last chunks) enc)]
      (conj handle-args data-arg))
    []))


src/clj_gearman/client.clj
(ns clj-gearman.client
  "Gearman client functions."
  (:require [clj-gearman.socket :as s]
            [clj-gearman.pool   :as p]
            [clj-gearman.header :as h]
            [clj-gearman.util   :as u]))


(defn read-response [socket]
  (let [opt (meta socket)
        error? #{"WORK_FAIL" "WORK_EXCEPTION" "ERROR"}
        done?  #{"WORK_COMPLETE"}
        chunk? #{"WORK_DATA"}
        warn?  #{"WORK_WARNING"}]

    (loop [result ""]
      (let [[code data] (s/response socket opt)]
        (when (warn? code) (.print *err* (last data)))
        (cond
          (error? code) [code data]
          (done?  code) [code (str result (last data))]
          :else
          (recur (if (chunk? code) (str result (last data)) result)))))))

(defn submit-job
  ([socket fn-name data] (submit-job socket fn-name data {}))

  ([socket fn-name data opt]
  (let [res (s/request socket
                     (:code opt (h/code "SUBMIT_JOB"))
                     fn-name
                     (:uniq opt (u/new-uniq))
                     data)]
    (if (= (res 0) "JOB_CREATED")
      (read-response socket)
      res))))


(defn submit-job-high
  ([socket fn-name data] (submit-job-high socket fn-name data {}))
  ([socket fn-name data opt]
   (submit-job socket fn-name data (assoc opt :code (h/code "SUBMIT_JOB_HIGH")))))


(defn submit-job-low
  ([socket fn-name data] (submit-job-low socket fn-name data {}))
  ([socket fn-name data opt]
   (submit-job socket fn-name data (assoc opt :code (h/code "SUBMIT_JOB_LOW")))))


(defn submit-job-bg
  ([socket fn-name data] (submit-job-bg socket fn-name data {}))
  ([socket fn-name data opt]
   (let [[code [job-handle] :as response] (s/request socket
                                                   (:code opt (h/code "SUBMIT_JOB_BG"))
                                                   fn-name
                                                   (:uniq opt (u/new-uniq))
                                                   data)]
     (if (= code "JOB_CREATED")
       [code job-handle]
       response))))

(defn submit-job-high-bg
  ([socket fn-name data] (submit-job-high-bg socket fn-name data {}))
  ([socket fn-name data opt]
   (submit-job-bg socket fn-name data (assoc opt :code (h/code "SUBMIT_JOB_HIGH_BG")))))

(defn submit-job-low-bg
  ([socket fn-name data] (submit-job-high-bg socket fn-name data {}))
  ([socket fn-name data opt]
   (submit-job-bg socket fn-name data (assoc opt :code (h/code "SUBMIT_JOB_LOW_BG")))))


(defn get-status [socket job-handle]
  (s/request socket (h/code "GET_STATUS") job-handle))

(defn get-status-unique [socket uniq]
  (s/request socket (h/code "GET_STATUS_UNIQUE") uniq))

(defn option-req [socket opt]
  (s/request socket (h/code "OPTION_REQ") opt))

(defn echo
  [socket data]
  (let [[code [response]] (s/request socket (h/code "ECHO_REQ") data)]
    [code response]))

(defn connect [client]
  (let [socket (p/connect-first (:job-servers client) client)]
    (when (:worker-exceptions client)
      (option-req socket "exceptions"))
    socket))



src/clj_gearman/socket.clj
(ns clj-gearman.socket
  "Low-level socket functions."
  (:import (java.net Socket ConnectException SocketTimeoutException)
           (java.io Closeable InputStream))
  (:require [clj-gearman.header :as h]
            [clj-gearman.util :as u]))

; We need to use a stupid wrapper in order to pile our
; client and worker metadata on the socket.
(deftype MetaSocket [^Socket socket _meta]

  clojure.lang.IObj
  (meta [_] _meta)
  (withMeta [_ m] (MetaSocket. socket m))

  clojure.lang.IDeref
  (deref [_] socket)

  java.io.Closeable
  (close [_] (.close socket))

  Object
  (toString [this]
    (str (.getName (class this))
         ": "
         (pr-str socket))))


(defn connect
  ([conn] (connect conn {} 0))
  ([{:keys [^String host ^Long port] :as conn} meta retry]
   (try
     (MetaSocket. (Socket. host port) meta)
     (catch ConnectException ex
       (if (pos? retry)
         (do (Thread/sleep 1000)
             (connect conn meta (dec retry)))
         (throw ex))))))

(defn disconnect [socket]
  (let [^Socket s @socket]
    (.close s)))

(defn read-bytes [^InputStream in ^Integer len]
  (if (pos? len)
    (byte-array (map (fn [_] (.read in)) (range len)))))

(defn read-int [in]
  (u/bytea->int (read-bytes in 4)))

(defn read-msg [socket enc]
  (let [^Socket s @socket
        r      (.getInputStream s)
        type   (read-bytes r 4)
        code   (read-int r)
        size   (read-int r)
        byte-a (read-bytes r size)]
    (list type code size (u/bytea->msg byte-a enc))))

(defn write-msg [socket type code enc msg]
  (let [^Socket s @socket
        w    (.getOutputStream s)
        data (u/msg->bytea msg enc)
        size (count data)]
    (.write w (u/concat-bytea
                [type
                 (u/int->bytea code)
                 (u/int->bytea size)
                 data]))
    (.flush w)))

(defn with-timeout [worker fun]
  ; Add some randomness to our wait period so multiple workers
  ; don't poll at the same time.
  (try
    (.setSoTimeout @worker (+ 5000 (rand-int 5000)))
    (fun)
    (catch SocketTimeoutException _
      ["" []])
    (finally
      (.setSoTimeout @worker 0))))

(defn response [socket opt]
  (let [[_ code _ data] (read-msg socket (:in-enc opt))]
    [(h/lookup code) data]))

(defn request [socket code & msg]
  (let [opt (meta socket)]
    (write-msg socket h/req code (:out-enc opt) msg)
    (response socket opt)))


src/clj_gearman/worker.clj
(ns clj-gearman.worker
  "Gearman worker functions."
  (:require [clj-gearman.socket :as s]
            [clj-gearman.header :as h]
            [clj-gearman.pool :as p]
            [clj-gearman.util :as u]))

(defn send-msg [socket code & args]
  (let [opt (meta socket)]
    (s/write-msg socket h/req (h/code code) (:out-enc opt) args)))

(defn can-do [socket fn-name]
  (send-msg socket "CAN_DO" fn-name))

(defn can-do-timeout [socket fn-name timeout]
  (send-msg socket "CAN_DO_TIMEOUT" fn-name timeout))

(defn cant-do [socket fn-name]
  (send-msg socket "CANT_DO" fn-name))

(defn reset-abilities [socket]
  (send-msg socket "RESET_ABILITIES"))

(defn pre-sleep [socket]
  (send-msg socket "PRE_SLEEP"))

(defn grab-job [socket]
  (s/request socket (h/code "GRAB_JOB")))

(defn grab-job-uniq [socket]
  (s/request socket (h/code "GRAB_JOB_UNIQ")))

(defn grab-job-all [socket]
  (s/request socket (h/code "GRAB_JOB_ALL")))

(defn work-data [socket job-handle data]
  (send-msg socket "WORK_DATA" job-handle data))

(defn work-warning [socket job-handle data]
  (send-msg socket "WORK_WARNING" job-handle data))

(defn work-status [socket job-handle numer denom]
  (send-msg socket "WORK_STATUS" job-handle numer denom))

(defn work-complete [socket job-handle data]
  (send-msg socket "WORK_COMPLETE" job-handle data))

(defn work-fail [socket job-handle]
  (send-msg socket "WORK_FAIL" job-handle))

(defn work-exception [socket job-handle exception]
  (send-msg socket "WORK_EXCEPTION" job-handle exception))

(defn set-client-id [socket uniq]
  (send-msg socket "SET_CLIENT_ID" uniq))


(defn run-task [socket job-handle fn-name workload]
  (let [fun (get-in (meta socket) [:can-do fn-name])]
    (if (= (u/fun-arity fun) 1)
      (try
        (work-complete socket job-handle (fun workload))
        (catch Throwable ex
          (work-exception socket job-handle ex)))
      (fun socket job-handle workload))))


(defn wait-for-task [socket]
  (s/with-timeout
    socket
    (fn []
      (let [[code data] (grab-job socket)]
        (if (= code "JOB_ASSIGN")
          [code data]
          (let [[code data] (s/request socket (h/code "PRE_SLEEP"))]
            (when (= code "NOOP")
              (grab-job socket))))))))

(defn work [socket]
  (let [[code [job-handle fn-name workload]] (wait-for-task socket)]
    (when (= code "JOB_ASSIGN")
      (if (contains? (:can-do (meta socket)) fn-name)
        (run-task socket job-handle fn-name workload)
        (do
          (work-fail socket job-handle)
          (cant-do socket fn-name))))))

(defn connect [worker]
  (let [socket (s/connect (first (:job-servers worker)) worker 10)]
    (doseq [ability (keys (get worker :can-do {}))]
      (can-do socket (name ability)))
    socket))

(defn pool
  "Create a pool of worker threads and start accepting tasks.
  Returns a function which will stop the pool when called."
  [worker]
  (p/worker-pool worker connect work))


src/clj_gearman/header.clj
(ns clj-gearman.header
  "Gearman request and response codes."
  (:require [clojure.set :refer [map-invert]]))


(def req (byte-array (map byte [0 \R \E \Q])))
(def res (byte-array (map byte [0 \R \E \S])))


(def code {
               "CAN_DO" 1   ;REQ    Worker
              "CANT_DO" 2   ;REQ    Worker
      "RESET_ABILITIES" 3   ;REQ    Worker
            "PRE_SLEEP" 4   ;REQ    Worker
             "(unused)" 5   ;-      -
                 "NOOP" 6   ;RES    Worker
           "SUBMIT_JOB" 7   ;REQ    Client
          "JOB_CREATED" 8   ;RES    Client
             "GRAB_JOB" 9   ;REQ    Worker
              "NO_JOB" 10   ;RES    Worker
          "JOB_ASSIGN" 11   ;RES    Worker
         "WORK_STATUS" 12   ;REQ    Worker
                            ;RES    Client
       "WORK_COMPLETE" 13   ;REQ    Worker
                            ;RES    Client
           "WORK_FAIL" 14   ;REQ    Worker
                            ;RES    Client
          "GET_STATUS" 15   ;REQ    Client
            "ECHO_REQ" 16   ;REQ    Client/Worker
            "ECHO_RES" 17   ;RES    Client/Worker
       "SUBMIT_JOB_BG" 18   ;REQ    Client
               "ERROR" 19   ;RES    Client/Worker
          "STATUS_RES" 20   ;RES    Client
     "SUBMIT_JOB_HIGH" 21   ;REQ    Client
       "SET_CLIENT_ID" 22   ;REQ    Worker
      "CAN_DO_TIMEOUT" 23   ;REQ    Worker
           "ALL_YOURS" 24   ;REQ    Worker
      "WORK_EXCEPTION" 25   ;REQ    Worker
                            ;RES    Client
          "OPTION_REQ" 26   ;REQ    Client/Worker
          "OPTION_RES" 27   ;RES    Client/Worker
           "WORK_DATA" 28   ;REQ    Worker
                            ;RES    Client
        "WORK_WARNING" 29   ;REQ    Worker
                            ;RES    Client
       "GRAB_JOB_UNIQ" 30   ;REQ    Worker
     "JOB_ASSIGN_UNIQ" 31   ;RES    Worker
  "SUBMIT_JOB_HIGH_BG" 32   ;REQ    Client
      "SUBMIT_JOB_LOW" 33   ;REQ    Client
   "SUBMIT_JOB_LOW_BG" 34   ;REQ    Client
    "SUBMIT_JOB_SCHED" 35   ;REQ    Client
    "SUBMIT_JOB_EPOCH" 36   ;REQ    Client
   "SUBMIT_REDUCE_JOB" 37   ;REQ    Client
  "SUBMIT_REDUCE_JOB_BACKGROUND"
                       38   ;REQ    Client
        "GRAB_JOB_ALL" 39   ;REQ    Worker
      "JOB_ASSIGN_ALL" 40   ;RES    Worker
   "GET_STATUS_UNIQUE" 41   ;REQ    Client
   "STATUS_RES_UNIQUE" 42   ;RES    Client
})

(def lookup (map-invert code))
